api/client/client.go

```diff
@@ -54,48 +52,116 @@ func init() {
 // Client is a gRPC Client that connects to a teleport auth server through TLS.
 type Client struct {
 	c Config
+
+	// tlsConfig is the *tls.Config for a successfully connected client.
+	tlsConfig *tls.Config
+
+	// dialer is the ContextDialer for a successfully connected client.
+	dialer ContextDialer
+
 	// grpc is the gRPC client specification for the auth server.
 	grpc proto.AuthServiceClient
+
 	// conn is a grpc connection to the auth server.
 	conn *grpc.ClientConn
+
 	// closedFlag is set to indicate that the services are closed.
 	closedFlag int32
 }
 
-// NewClient returns a new auth client that uses mutual TLS authentication and
-// connects to the remote server using the Dialer or Addrs given in Config.
-func NewClient(cfg Config) (*Client, error) {
+// New creates a new API client with a connection to a Teleport server. The connection is
+// formed using the Dialer and CredentialsProviders given in config. If Addrs are provided, a
+// basic Dialer will be created. If Credentials are provided, a basic CredentialsProvider will
+// will be added to the list of CredentialsProviders.
+func New(cfg Config) (*Client, error) {
 	if err := cfg.CheckAndSetDefaults(); err != nil {
 		return nil, trace.Wrap(err)
 	}
 
 	c := &Client{c: cfg}
-	dialer := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
-		if c.isClosed() {
-			return nil, trace.ConnectionProblem(nil, "client is closed")
+	if err := c.connect(cfg.NoPingCheck); err != nil {
```

Comment:
Why do you need to pass this into `c.connect()` explicitly if it's a part of config the client is initialized with anyway?